internal/build/resolver/resolver.go

```diff
@@ -0,0 +1,60 @@
+package resolver
+
+import (
+	"context"
+	"errors"
+
+	"github.com/buildkite/cli/v3/internal/build"
+	"github.com/buildkite/cli/v3/internal/pipeline"
+)
+
+var resolved bool
+var resolvedBuild *build.Build
+var resolvedError error
+
+// BuildResolverFn is a function for finding a build. It returns an error if an irrecoverable scenario happens and
+// should halt execution. Otherwise, if the resolver does not find a build, it should return (nil, nil) to indicate
+// this. ie. no error occurred, but no build was found either
+//
+// Some build resolvers may need to limit scope to a pipeline so that is provided to the resolver at runtime
+type BuildResolverFn func(context.Context, pipeline.Pipeline) (*build.Build, error)
```

Comment:
Do we like the ergonomics of this? Its similar to the pipeline resolver but we should start passing around context (for use in things like API requests). And some build resolvers will need to know the resolved pipeline, should that be passed in here or passed by closure like the other pipeline resolver functions?

Reply:
Or building on the comment below, this could accept a PipelineResolverFn that is cached instead: 
```suggestion
type BuildResolverFn func(context.Context, pipelineresolver.PipelineResolverFn) (*build.Build, error)
```

Reply:
I opted to remove this entirely from the definition of a resolver. If a resolver has another kind of dependency, such as needing to resolve a pipeline, it should be provided by binding it to a closure instead